### PR TITLE
[clang][analyzer] Bring cplusplus.ArrayDelete out of alpha

### PR DESCRIPTION
**Pull request by Discookie**
The checker finds a type of undefined behavior, where if the type of a pointer to an object-array is different from the objects' underlying type, calling `delete[]` is undefined, as the size of the two objects might be different.

The checker has been in alpha for a while now, it is a simple checker that causes no crashes, and considering the severity of the issue, it has a low result-count on open-source projects (in my last test-run on my usual projects, it had 0 results).

This commit cleans up the documentation and adds docs for the limitation related to tracking through references, in addition to moving it to `cplusplus`.

### CODE REVIEW

**Review by NagyDonat — COMMENTED** [review]
LGTM. I think you can merge this in a few days (if nobody requests changes).

**Comment by steakhal** [comment]
I'm fine with the change.

File: clang/docs/analyzer/checkers.rst [hunk 1]
```diff
@@ -340,6 +340,50 @@ cplusplus
 
 C++ Checkers.
 
+.. _cplusplus-ArrayDelete:
+
+cplusplus.ArrayDelete (C++)
+"""""""""""""""""""""""""""
+Reports destructions of arrays of polymorphic objects that are destructed as
```
Comment (whisperity) [inline clang/docs/analyzer/checkers.rst:347]:
```suggestion
"""""""""""""""""""""""""""

Reports destructions of arrays of polymorphic objects that are destructed as
```

Style nit.

File: clang/docs/analyzer/checkers.rst [hunk 2]
```diff
@@ -340,6 +340,50 @@ cplusplus
...
+their base class. If the dynamic type of the array's object is different from
+its static type, calling `delete[]` is undefined.
```
Comment (whisperity) [inline clang/docs/analyzer/checkers.rst:349]:
```suggestion
their base class. If the dynamic type of the array is different from
its static type, calling `delete[]` is undefined.
```

File: clang/docs/analyzer/checkers.rst [hunk 3]
```diff
@@ -340,6 +340,50 @@ cplusplus
...
+
+This checker corresponds to the CERT rule `EXP51-CPP: Do not delete an array through a pointer of the incorrect type <https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP51-CPP.+Do+not+delete+an+array+through+a+pointer+of+the+incorrect+type>`_.
```
Comment (whisperity) [inline clang/docs/analyzer/checkers.rst:351]:
```suggestion
This checker corresponds to the SEI CERT rule `EXP51-CPP: Do not delete an array through a pointer of the incorrect type <https://wiki.sei.cmu.edu/confluence/display/cplusplus/EXP51-CPP.+Do+not+delete+an+array+through+a+pointer+of+the+incorrect+type>`_.
```

Style nit, to toe the line.

File: clang/docs/analyzer/checkers.rst [hunk 4]
```diff
@@ -340,6 +340,50 @@ cplusplus
...
+   Base *x = new Derived[10]; // note: Casting from 'Derived' to 'Base' here
+   return x;
+ }
+
+ void foo() {
+   Base *x = create();
+   delete[] x; // warn: Deleting an array of 'Derived' objects as their base class 'Base' is undefined
+ }
+
+**Limitations**
```
Comment (whisperity) [inline clang/docs/analyzer/checkers.rst:371]: (Shouldn't this be a sub-heading?)

File: clang/lib/StaticAnalyzer/Checkers/CXXDeleteChecker.cpp [hunk 5]
```diff
@@ -220,11 +220,11 @@ CXXDeleteChecker::PtrCastVisitor::VisitNode(const ExplodedNode *N,
                                                     /*addPosRange=*/true);
 }
 
-void ento::registerCXXArrayDeleteChecker(CheckerManager &mgr) {
+void ento::registerArrayDeleteChecker(CheckerManager &mgr) {
```
Comment (whisperity) [inline clang/lib/StaticAnalyzer/Checkers/CXXDeleteChecker.cpp:223]: Why was the `CXX` removed here if the checker is still named `cplusplus.ArrayDelete`.

**Review by whisperity — COMMENTED** [review]
Style nits only.

File: clang/lib/StaticAnalyzer/Checkers/CXXDeleteChecker.cpp (hunk 5 above)
Comment (Discookie) [reply clang/lib/StaticAnalyzer/Checkers/CXXDeleteChecker.cpp:223]: Brings it more in line with similar C++ checkers, see eg. `DeleteWithNonVirtualDtorChecker` in the same file.

The only place where this name matters is the `Checkers.td` where it is under the `cplusplus` group anyways.

File: clang/docs/analyzer/checkers.rst (hunk 4 above)
Comment (NagyDonat) [reply clang/docs/analyzer/checkers.rst:371]: No, this is the formatting that's consistently used in this document.

**Review by NagyDonat — APPROVED** [review]
Still LGTM, apply the few suggested changes and then feel free to merge the commit.

**Pull request merged**